File: assembler/assembler.lua

--[[
	Author: Luis Vijande
	Date: 18/10/2023

	Supposedly this should work with all addressings
	in hex using $numbers
	in bin using %numbers
	in decimal using numbers

	All 54 official instructions should work
	It evaluates line by line
	If you write a number by itself it will add it to the assembly as the number 
]]--

Instructions = {
	ADC = {
		inx = 0x61,
		zpi = 0x65,
		imm = 0x69,
		abs = 0x6d,
		iny = 0x71,
		zpx = 0x75,
		aby = 0x79,
		abx = 0x7d,
	},
	AND = {
		inx = 0x21,
		zpi = 0x25,
		imm = 0x29,
		abs = 0x2d,
		iny = 0x31,
		zpx = 0x35,
		aby = 0x39,
		abx = 0x3d,
	},
	ASL = {
		zpi = 0x06,
		acc = 0x0a,
		abs = 0x0e,
		zpx = 0x16,
		abx = 0x1e,
	},
	BCC = {
		rel = 0x90,
	},
	BCS = {
		rel = 0xb0,
	}, 
	BEQ = {
		rel = 0xf0,
	},
	BIT = {
		zpi = 0x24,
		abs = 0x2c,
	},
	BMI = {
		rel = 0x30,
	},
	BNE = {
		rel = 0xd0,
	},
	BPL = {
		rel = 0x10,
	},
	BRK = {
		imp = 0x00,
	},
	BVC = {
		rel = 0x50,
	},
	BVS = {
		rel = 0x70,
	},
	CLC = {
		imp = 0x18,
	},
	CLD = {
		imp = 0xd8,
	},
	CLI = {
		imp = 0x58,
	},
	CLV = {
		imp = 0xb8
	},
	CMP = {
		inx = 0xc1,
		zpi = 0xc5,
		imm = 0xc9,
		abs = 0xcd,
		iny = 0xd1,
		zpx = 0xd5,
		aby = 0xd9,
		abx = 0xdd,
	},
	CPX = {
		imm = 0xe0,
		zpi = 0xe4,
		abs = 0xec,
	},
	CPY = {
		imm = 0xc0,
		zpi = 0xc4,
		abs = 0xcc,
	},
	DEC = {
		zpi = 0xc6,
		zpx = 0xd6,
		abs = 0xce,
		abx = 0xde,
	},
	DEX = {
		imp = 0xca,
	},
	DEY = {
		imp = 0x88,
	},
	EOR = {
		inx = 0x41,
		zpi = 0x45,
		imm = 0x49,
		abs = 0x4d,
		iny = 0x51,
		zpx = 0x55,
		aby = 0x59,
		abx = 0x5d,
	},
	INC = {
		zpi = 0xe6,
		zpx = 0xf6,
		abs = 0xee,
		abx = 0xfe,
	},
	INX = {
		imp = 0xe8,
	},
	INY = {
		imp = 0xc8,
	},
	JMP = {
		abs = 0x4c,
		ind = 0x6c,
	},
	JSR = {
		abs = 0x20,
	},
	LDA = {
		inx = 0xa1,
		zpi = 0xa5,
		imm = 0xa9,
		abs = 0xad,
		iny = 0xb1,
		zpx = 0xb5,
		aby = 0xb9,
		abx = 0xbd,
	},
	LDX = {
		imm = 0xa2,
		zpi = 0xa6,
		zpy = 0xb6,
		abs = 0xae,
		aby = 0xbe,
	},
	LDY = {
		imm = 0xa0,
		zpi = 0xa4,
		zpx = 0xb4,
		abs = 0xac,
		abx = 0xbc,
	},
	LSR = {
		acc = 0x4a,
		zpi = 0x46,
		zpx = 0x56,
		abs = 0x4e,
		abx = 0x5e,
	},
	NOP = {
		imp = 0xea,
	},
	ORA = {
		inx = 0x01,
		zpi = 0x05,
		imm = 0x09,
		abs = 0x0d,
		iny = 0x11,
		zpx = 0x15,
		aby = 0x19,
		abx = 0x1d,
	},
	PHA = {
		imp = 0x48,
	},
	PHP = {
		imp = 0x08,
	},
	PLA = {
		imp = 0x68,
	},
	PLP = {
		imp = 0x28,
	},
	ROL = {
		acc = 0x2a,
		zpi = 0x26,
		zpx = 0x36,
		abs = 0x2e,
		abx = 0x3e,
	},
	ROR = {
		acc = 0x6a,
		zpi = 0x66,
		zpx = 0x76,
		abs = 0x6e,
		abx = 0x7e,
	},
	RTI = {
		imp = 0x40,
	},
	RTS = {
		imp = 0x60,
	},
	SBC = {
		inx = 0xe1,
		zpi = 0xe5,
		imm = 0xe9,
		abs = 0xed,
		iny = 0xf1,
		zpx = 0xf5,
		aby = 0xf9,
		abx = 0xfd,
	},
	SEC = {
		imp = 0x38,
	},
	SED = {
		imp = 0xf8,
	},
	SEI = {
		imp = 0x78,
	},
	STA = {
		inx = 0x81,
		zpi = 0x85,
		abs = 0x8d,
		iny = 0x91,
		zpx = 0x95,
		aby = 0x99,
		abx = 0x9d,
	},
	STX = {
		zpi = 0x86,
		zpy = 0x96,
		abs = 0x8e,
	},
	STY = {
		zpi = 0x84,
		zpx = 0x94,
		abs = 0x8c,
	},
	TAX = {
		imp = 0xaa,
	},
	TAY = {
		imp = 0xa8,
	},
	TSX = {
		imp = 0xba,
	},
	TXA = {
		imp = 0x8a,
	},
	TXS = {
		imp = 0x9a,
	},
	TYA = {
		imp = 0x98,
	},
}

function GetAddressing(text)
	local res ={}

	if text == nil then
		res['addr'] = 'abs'

	elseif text:match '^%*[-+]%d?%d$' then
		res['addr'] = 'rel'
		res['out'] = tonumber(text:match '([+-]%d+)')

	elseif text:match '^$%x+$' then
		local num = tonumber(text:match '(%x+)', 16)
		if num < 256 then
			res['addr'] = 'zpi'
			res['out'] = num
		else
			res['addr'] = 'abs'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%d+$' then
		local num = tonumber(text:match '(%d+)')
		if num < 256 then
			res['addr'] = 'zpi'
			res['out'] = num
		else
			res['addr'] = 'abs'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%%[01]+$' then
		local num = tonumber(text:match '([01]+)', 2)
		if num < 256 then
			res['addr'] = 'zpi'
			res['out'] = num
		else
			res['addr'] = 'abs'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end

	elseif text:match '^$%x+,%s*,%s*x$' then
		local num = tonumber(text:match '(%x+)', 16)
		if num < 256 then
			res['addr'] = 'zpx'
			res['out'] = num
		else
			res['addr'] = 'abx'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%d+%s*,%s*x$' then
		local num = tonumber(text:match '(%d+)')
		if num < 256 then
			res['addr'] = 'zpx'
			res['out'] = num
		else
			res['addr'] = 'abx'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%%[01]+%s*,%s*x$' then
		local num = tonumber(text:match '([01]+)', 2)
		if num < 256 then
			res['addr'] = 'zpx'
			res['out'] = num
		else
			res['addr'] = 'abx'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end

	elseif text:match '^$%x+,%s*,%s*y$' then
		local num = tonumber(text:match '(%x+)', 16)
		if num < 256 then
			res['addr'] = 'zpy'
			res['out'] = num
		else
			res['addr'] = 'aby'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%d+%s*,%s*y$' then
		local num = tonumber(text:match '(%d+)')
		if num < 256 then
			res['addr'] = 'zpy'
			res['out'] = num
		else
			res['addr'] = 'aby'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end
	elseif text:match '^%%[01]+%s*,%s*y$' then
		local num = tonumber(text:match '([01]+)', 2)
		if num < 256 then
			res['addr'] = 'zpy'
			res['out'] = num
		else
			res['addr'] = 'aby'
			res['out'] = {}
			res['out'][1] = num & 0x00ff
			res['out'][2] = (num & 0xff00) >> 8
		end

	elseif text:match '^%(%s*$%x+%s*,%s*x%s*%)$' then
		res['addr'] = 'inx'
		res['out'] = tonumber(text:match '(%x+)', 16)
	elseif text:match '^%(%s*%d+%s*,%s*x%s*%)$' then
		res['addr'] = 'inx'
		res['out'] = tonumber(text:match '(%d+)')
	elseif text:match '^%(%s*%%[01]+%s*,%s*x%s*%)$' then
		res['addr'] = 'inx'
		res['out'] = tonumber(text:match '([01]+)', 2)

	elseif text:match '^%(%s*$%x+%s*%)%s*,%s*y$' then
		res['addr'] = 'iny'
		res['out'] = tonumber(text:match '(%x+)', 16)
	elseif text:match '^%(%s*%d+%s*%)%s*,%s*y$' then
		res['addr'] = 'iny'
		res['out'] = tonumber(text:match '(%d+)')
	elseif text:match '^%(%s*%%[01]%s*%)%s*,%s*y$' then
		res['addr'] = 'iny'
		res['out'] = tonumber(text:match '([01]+)', 2)

	elseif text:match '^%(%s*$%x+%s*%)$' then
		local num = tonumber(text:match '(%x+)', 16)
		res['addr'] = 'ind'
		res['out'] = {}
		res['out'][1] = num & 0x00ff
		res['out'][2] = (num & 0xff00) >> 8
	elseif text:match '^%(%s*%d+%s*%)$' then
		local num = tonumber(text:match '(%d+)')
		res['addr'] = 'ind'
		res['out'] = {}
		res['out'][1] = num & 0x00ff
		res['out'][2] = (num & 0xff00) >> 8
	elseif text:match '^%(%s*%%[01]+%s*%)$' then
		local num = tonumber(text:match '([01]+)', 2)
		res['addr'] = 'ind'
		res['out'] = {}
		res['out'][1] = num & 0x00ff
		res['out'][2] = (num & 0xff00) >> 8


	elseif text:match '^#$%x+$' then
		res['addr'] = 'imm'
		res['out'] = tonumber(text:match '(%x+)', 16)
	elseif text:match '^#%d+$' then
		res['addr'] = 'imm'
		res['out'] = tonumber(text:match '(%d+)')
	elseif text:match '^#%%[01]+$' then
		res['addr'] = 'imm'
		res['out'] = tonumber(text:match '([01]+)', 2)

	elseif text:match '^[Aa]$' then
		res['addr'] = 'acc'
	end

	return res
end

function AssembleLine(line)
	if line == '' or line == nil then
		return nil
	end
	local assembly = {}
	if line:find '%a%a%a' then

		local instr = string.upper(line:match '^%s*(%a%a%a)')
		local addr  = line:match '^%s*%a%a%a%s+(.+)%s*$'
		local addrType = GetAddressing(addr)
		local an = addrType['addr']
		local out = addrType['out']
		-- check if instruction
		if instr:match('%a%a%a') then
			if Instructions[instr] ~= nil and Instructions[instr][an] ~= nil then
				table.insert(assembly, Instructions[instr][an])
				-- do nothing
				if an == 'imp' or an == 'acc' then
				-- 8-bit instructions
				elseif an == 'imm' or an == 'zpi' or an == 'zpx' or an == 'zpy' or an == 'rel' or an == 'iny' or an == 'inx' then
					table.insert(assembly, out)
				-- 16-bit instructions
				elseif (an == 'abs' or an == 'abx' or an == 'aby' or an == 'ind' ) and type(out) == 'table' then
					table.insert(assembly, out[1])
					table.insert(assembly, out[2])
				end
			else
				error('Invalid operation')
			end
		end
	else
		if line:match '^%s*$%x+$%s*$' then
			table.insert(assembly, tonumber(line:match '(%x+)', 16))
		elseif line:match '^%s*%d+%s*$' then
			table.insert(assembly, tonumber(line:match '(%d+)'))
		elseif line:match '^%s*%%[01]+%s*$' then
			table.insert(assembly, tonumber(line:match '([01]+)', 2))
		end
	end
	return assembly
end

function Assemble(code)
	local assembly = {}
	for line in code:gmatch '([^\n]+)' do
		local res = AssembleLine(line)
		if res ~= nil then
			for _, i in ipairs(res) do
				table.insert(assembly, i)
			end
		end
	end
	return assembly
end

File: NesEmu/src/utils/Assembler.cpp
#include "pch.hpp"
#include "Assembler.hpp"

#include <ctre/ctre.hpp>

#include <print>

#include <lud_parser.hpp>


namespace A6502
{

static std::string upper(std::string s);

Assembler& Assembler::Assemble(const std::string& code)
{

	u16 write_pos = 0;

	// split lines
	for (const auto& line : ctre::split <"\n">(code))
	{
		std::string s = upper(line.str());
		if (auto m = ctre::search<"@RESET\\s+\\$([A-F0-9]{4})">(s))
		{
			u16 dat = Lud::parse_num<u16>(m.get<1>().view(), 16);
			u8 lo = dat & 0x00FF;
			u8 hi = ( dat & 0xFF00 ) >> 8;
			m_bus->Write(0xFFFC, lo);
			m_bus->Write(0xFFFD, hi);
		}
		else if (auto m = ctre::search<"@AT\\s+\\$([A-F0-9]{4})">(s))
		{
			write_pos = Lud::parse_num<u16>(m.get<1>().view(), 16);
		}
		else if (auto m = ctre::search<"([A-Z]{3})(\\s+([0-9A-FXY,\\-()#\\$]+))?">(s))
		{
			Opcode op = {};
			// parse line
			op.instruction = GetInstructionName(m.get<1>());
			auto [mode, dat] = ParseAddressingMode(m.get<3>());
			op.mode = mode;
			auto bytes = GetBytesForAddressingMode(mode);

			if (auto it = std::find(s_instructions.begin(), s_instructions.end(), op); it != s_instructions.end())
			{
				u8 index = static_cast<u8>( it - s_instructions.begin() );
				switch (bytes)
				{
				case 3:
				{
					u8 hi = ( dat & 0xFF00 ) >> 8;
					m_bus->Write(write_pos + 2, hi);
					[[fallthrough]];
				}
				case 2:
				{
					if (op.mode == AddressingModeName::REL)
					{
						i8 lo = static_cast<i8>( dat );
						m_bus->Write(write_pos + 1, lo);
					}
					else
					{
						u8 lo = dat & 0x00FF;
						m_bus->Write(write_pos + 1, lo);
					}
					[[fallthrough]];
				}
				case 1:
					m_bus->Write(write_pos, index);
				}
				write_pos += bytes;
			}
			else
			{
				std::println("{:s} this opcode does not exist, ignoring", s);
			}
		}
	}
	return *this;
}

void Assembler::Clean()
{
}

InstructionName Assembler::GetInstructionName(const std::string_view name)
{
	if (name == "ADC") return A6502::InstructionName::ADC;
	if (name == "AND") return A6502::InstructionName::AND;
	if (name == "ASL") return A6502::InstructionName::ASL;
	if (name == "BCC") return A6502::InstructionName::BCC;
	if (name == "BCS") return A6502::InstructionName::BCS;
	if (name == "BEQ") return A6502::InstructionName::BEQ;
	if (name == "BIT") return A6502::InstructionName::BIT;
	if (name == "BMI") return A6502::InstructionName::BMI;
	if (name == "BNE") return A6502::InstructionName::BNE;
	if (name == "BPL") return A6502::InstructionName::BPL;
	if (name == "BRK") return A6502::InstructionName::BRK;
	if (name == "BVC") return A6502::InstructionName::BVC;
	if (name == "BVS") return A6502::InstructionName::BVS;
	if (name == "CLC") return A6502::InstructionName::CLC;
	if (name == "CLD") return A6502::InstructionName::CLD;
	if (name == "CLI") return A6502::InstructionName::CLI;
	if (name == "CLV") return A6502::InstructionName::CLV;
	if (name == "CMP") return A6502::InstructionName::CMP;
	if (name == "CPX") return A6502::InstructionName::CPX;
	if (name == "CPY") return A6502::InstructionName::CPY;
	if (name == "DEC") return A6502::InstructionName::DEC;
	if (name == "DEX") return A6502::InstructionName::DEX;
	if (name == "DEY") return A6502::InstructionName::DEY;
	if (name == "EOR") return A6502::InstructionName::EOR;
	if (name == "INC") return A6502::InstructionName::INC;
	if (name == "INX") return A6502::InstructionName::INX;
	if (name == "INY") return A6502::InstructionName::INY;
	if (name == "JMP") return A6502::InstructionName::JMP;
	if (name == "JSR") return A6502::InstructionName::JSR;
	if (name == "LDA") return A6502::InstructionName::LDA;
	if (name == "LDX") return A6502::InstructionName::LDX;
	if (name == "LDY") return A6502::InstructionName::LDY;
	if (name == "LSR") return A6502::InstructionName::LSR;
	if (name == "NOP") return A6502::InstructionName::NOP;
	if (name == "ORA") return A6502::InstructionName::ORA;
	if (name == "PHA") return A6502::InstructionName::PHA;
	if (name == "PHP") return A6502::InstructionName::PHP;
	if (name == "PLA") return A6502::InstructionName::PLA;
	if (name == "PLP") return A6502::InstructionName::PLP;
	if (name == "ROL") return A6502::InstructionName::ROL;
	if (name == "ROR") return A6502::InstructionName::ROR;
	if (name == "RTI") return A6502::InstructionName::RTI;
	if (name == "RTS") return A6502::InstructionName::RTS;
	if (name == "SBC") return A6502::InstructionName::SBC;
	if (name == "SEC") return A6502::InstructionName::SEC;
	if (name == "SED") return A6502::InstructionName::SED;
	if (name == "SEI") return A6502::InstructionName::SEI;
	if (name == "STA") return A6502::InstructionName::STA;
	if (name == "STX") return A6502::InstructionName::STX;
	if (name == "STY") return A6502::InstructionName::STY;
	if (name == "TAX") return A6502::InstructionName::TAX;
	if (name == "TAY") return A6502::InstructionName::TAY;
	if (name == "TSX") return A6502::InstructionName::TSX;
	if (name == "TXA") return A6502::InstructionName::TXA;
	if (name == "TXS") return A6502::InstructionName::TXS;
	if (name == "TYA") return A6502::InstructionName::TYA;
	if (name == "ALR") return A6502::InstructionName::ALR;
	if (name == "ANC") return A6502::InstructionName::ANC;
	if (name == "ARR") return A6502::InstructionName::ARR;
	if (name == "AXS") return A6502::InstructionName::AXS;
	if (name == "LAX") return A6502::InstructionName::LAX;
	if (name == "SAX") return A6502::InstructionName::SAX;
	if (name == "DCP") return A6502::InstructionName::DCP;
	if (name == "ISC") return A6502::InstructionName::ISC;
	if (name == "RLA") return A6502::InstructionName::RLA;
	if (name == "RRA") return A6502::InstructionName::RRA;
	if (name == "SLO") return A6502::InstructionName::SLO;
	if (name == "SRE") return A6502::InstructionName::SRE;
	if (name == "DOP") return A6502::InstructionName::DOP;
	if (name == "TOP") return A6502::InstructionName::TOP;
	if (name == "STP") return A6502::InstructionName::STP;
	if (name == "XAA") return A6502::InstructionName::XAA;
	if (name == "AHX") return A6502::InstructionName::AHX;
	if (name == "TAS") return A6502::InstructionName::TAS;
	if (name == "SHY") return A6502::InstructionName::SHY;
	if (name == "SHX") return A6502::InstructionName::SHX;
	if (name == "LAS") return A6502::InstructionName::LAS;

	std::println("Instruction {:s} is unknow", name);
	return A6502::InstructionName::___;
}

AddressingModeData Assembler::ParseAddressingMode(const std::string_view mode)
{
	if (auto m = ctre::match<"">(mode))
	{
		return { A6502::AddressingModeName::IMP, 0 };
	}
	if (auto m = ctre::match<"A">(mode))
	{
		return { A6502::AddressingModeName::ACC, 0 };
	}
	if (auto m = ctre::match<"#\\$([A-F0-9]{2})">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::IMM, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{2})">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ZPI, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{2}),X">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ZPX, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{2}),Y">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ZPY, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"(-?[0-9]{1,3})">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::REL, static_cast<u16>( Lud::parse_num<i8>(dat.view(), 10) ) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{4})">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ABS, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{4}),X">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ABX, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\$([A-F0-9]{4}),Y">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::ABY, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\(\\$([A-F0-9]{4})\\)">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::IND, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\(\\$([A-F0-9]{4}),X\\)">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::INX, Lud::parse_num<u16>(dat.view(), 16) };
	}
	if (auto m = ctre::match<"\\(\\$([A-F0-9]{4})\\),Y">(mode))
	{
		auto dat = m.get<1>();
		return { A6502::AddressingModeName::INY, Lud::parse_num<u16>(dat.view(), 16) };
	}
	std::println("Unrecognized addressing mode {:s}", mode);
	return { A6502::AddressingModeName::___, 0 };
}

std::string upper(std::string s)
{
	std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c)
		{
			return std::toupper(c);
		});
	return s;
}

}